### ISSUE TITLE
ImageFromDockerfileBuilder can't find Dockerfile

### ISSUE THREAD

**Issue by thundersen** [testcontainers/testcontainers-dotnet#405]
**Describe the bug**
I'm trying to build an image from a Dockerfile that's not in my test working directory:

```
var forwarderDir = "../../../../forwarder";
Console.WriteLine(string.Join("\n", Directory.GetFiles(forwarderDir)));
// > ../../../../forwarder/Dockerfile

var forwarderImageBuilder = new ImageFromDockerfileBuilder()
    .WithDockerfileDirectory(forwarderDir)
    .WithName("forwarder_integrationtest")
    .WithDeleteIfExists(true);

var forwarderImage = await forwarderImageBuilder.Build();
```
I get an exception saying that Dockerfile can't be found:
```
Docker.DotNet.DockerApiException : Docker API responded with status code=InternalServerError, response={"message":"Cannot locate specified Dockerfile: Dockerfile"}

    Stack Trace:
       at Docker.DotNet.DockerClient.HandleIfErrorResponseAsync(HttpStatusCode statusCode, HttpResponseMessage response, IEnumerable`1 handlers)
   at Docker.DotNet.DockerClient.MakeRequestForStreamAsync(IEnumerable`1 errorHandlers, HttpMethod method, String path, IQueryString queryString, IRequestContent body, IDictionary`2 headers, TimeSpan timeout, CancellationToken token)
   at DotNet.Testcontainers.Clients.DockerImageOperations.BuildAsync(IImageFromDockerfileConfiguration config, CancellationToken ct)
   at DotNet.Testcontainers.Images.Builders.ImageFromDockerfileBuilder.Build()
   at VeVantZeData.Collector.Integration.InfluxDBIntegrationTests.CanWriteToInfluxDB() in /home/alex/code/personal/timberborn/vevantzedata/collector_test/Integration/InfluxDBIntegrationTests.cs:line 50
   at NUnit.Framework.Internal.TaskAwaitAdapter.GenericAdapter`1.BlockUntilCompleted()
   at NUnit.Framework.Internal.MessagePumpStrategy.NoMessagePumpStrategy.WaitForCompletion(AwaitAdapter awaiter)
   at NUnit.Framework.Internal.AsyncToSyncAdapter.Await(Func`1 invoke)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunTestMethod(TestExecutionContext context)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.Execute(TestExecutionContext context)
   at NUnit.Framework.Internal.Execution.SimpleWorkItem.<>c__DisplayClass4_0.<PerformWork>b__0()
   at NUnit.Framework.Internal.ContextUtils.<>c__DisplayClass1_0`1.<DoIsolated>b__0(Object _)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at NUnit.Framework.Internal.ContextUtils.DoIsolated(ContextCallback callback, Object state)
   at NUnit.Framework.Internal.ContextUtils.DoIsolated[T](Func`1 func)
   at NUnit.Framework.Internal.Execution.SimpleWorkItem.PerformWork()
```

**Expected behavior**
Image is built with no exception.

Ubuntu 20.04 - Linux 5.4.0-88-generic
.Net 5.0.402
DotNet.Testcontainers 1.5.0

**Comment by HofmeisterAn** [comment]
`WithDockerfileDirectory` expects the directory of the `Dockerfile`, not the `Dockerfile` path. `../../../../forwarder` should work. See the corresponding unit test:
https://github.com/HofmeisterAn/dotnet-testcontainers/blob/0b37e212871e81f3af808ee0d48fe4cdbf3f5340/tests/DotNet.Testcontainers.Tests/Unit/Images/ImageFromDockerfileTest.cs#L44

**Comment by thundersen** [comment]
Hey @HofmeisterAn, thanks a lot for the quick reply (and working on this library in the first place)!

> `../../../../forwarder` should work. 

But that's exactly what I'm doing, isn't it? I'm supplying the relative path to the directory that contains the `Dockerfile`.

**Comment by HofmeisterAn** [comment]
> But that's exactly what I'm doing, isn't it?

Ah, I see. I was probably confused by `// > ../../../../forwarder/Dockerfile`.

It looks like Testcontainers finds the `Dockerfile`, otherwise it will throw:
https://github.com/HofmeisterAn/dotnet-testcontainers/blob/ef67133a1287a5cd183af9ae21e56fd44089f69e/src/DotNet.Testcontainers/Images/DockerfileArchive.cs#L54

Can you test what happens with an absolute path?

**Comment by thundersen** [comment]
@HofmeisterAn 

> Ah, I see. I was probably confused by // > ../../../../forwarder/Dockerfile.

Sorry. I included that as a comment to indicate that the file is listed by `Console.WriteLine(string.Join("\n", Directory.GetFiles(forwarderDir)));`

> Can you test what happens with an absolute path?

Same result.

**Comment by HofmeisterAn** [comment]
Ok. Can you please double check if the path is correct. It's working fine here:

```csharp
const string dockerfileDirectory = "../../../../../tests/DotNet.Testcontainers.Tests/Assets";
var directoryExists = Directory.Exists(dockerfileDirectory);

var imageFromDockerfileBuilder = new ImageFromDockerfileBuilder()
  .WithName("alpine:custom")
  .WithDockerfileDirectory(dockerfileDirectory)
  .WithDeleteIfExists(true);
```

creates

```
REPOSITORY                        TAG       IMAGE ID       CREATED        SIZE
alpine                            custom    20507ca03e33   3 hours ago    5.6MB
```

Can you use a simple Dockerfile too? Like the example in the unit test?

**Comment by thundersen** [comment]
I am certain that the path is correct, because I verified that it contains the `Dockerfile` with `Console.WriteLine(string.Join("\n", Directory.GetFiles(forwarderDir)));`

I tried placing the Dockerfile in the same directory as the test and weirdly building the image seems to work. I only changed the `forwarderDir` from `../../../../forwarder` to `../../..`.

That would be a workaround I could live with, but now it crashes when trying to start the container with `Docker.DotNet.DockerApiException : Docker API responded with status code=NotFound, response={"message":"pull access denied for forwarder_integrationtest, repository does not exist or may require 'docker login': denied: requested access to the resource is denied"}`

My adapted `Dockerfile` looks like this:
```
# syntax=docker/dockerfile:1
FROM mcr.microsoft.com/dotnet/sdk:5.0 AS build-env
WORKDIR /app

# Copy csproj and restore as distinct layers
COPY ../forwarder/*.csproj ./
RUN dotnet restore

# Copy everything else and build
COPY ../forwarder/. ./
RUN dotnet publish -c Release -o out

# Build runtime image
FROM mcr.microsoft.com/dotnet/aspnet:5.0
WORKDIR /app
COPY --from=build-env /app/out .
ENTRYPOINT ["dotnet", "Forwarder.dll"]
```

**Comment by HofmeisterAn** [comment]
Your `Dockerfile` won't work. The Docker Remote API requires a tar archive that includes all assets. .NET Testcontainers uses the `Dockerfile` directory to create the tar archive. All directories and files in the `Dockerfile` directory are included (except directories and files in `.dockerignore`). Parent directories and files **are not** included. See `Add` and `Copy` in the Docker [documentation](https://docs.docker.com/engine/reference/builder/):

> The `<src>` path must be inside the context of the build; you cannot `ADD ../something /something`, because the first step of a `docker build` is to send the context directory (and subdirectories) to the docker daemon.

You need to restructure your project. It's probably easier to put the `Dockerfile` to the root level and ignore unnecessary files with `.dockerignore`.

**Comment by thundersen** [comment]
> Parent directories and files are not included.

Right. Completely forgot about that. But that was just a test to see, if the image would be built (which it was).
It would also build, if I copied the `forwarder` directory into the tests' `bin` directory manually before test execution, but would fail with the same `DockerApiException` as in my initial post if I copied it automatically using a [`OneTimeSetUp`](https://docs.nunit.org/articles/nunit/writing-tests/attributes/onetimesetup.html) annotated setup method which did the copying before any of the testcontainers code would run.

Anyway I decided to work around this by building the image using plain `docker build` in an `AfterBuild` target in my `csproj`:
```
<Target Name="CustomAfterBuild" AfterTargets="Build" >
    <Exec Command="docker build -t forwarder_integrationtest ../forwarder" Condition=" '$(OS)' == 'Unix' " />
</Target>
```

Thank you again for your help @HofmeisterAn! I wish I could have gotten to the bottom of why this wouldn't work as intended, but I'm happy with my workaround and will close this issue now, because I've already spent way too much time on this. If I run into this again and find out more in the future, I will let you know.
